### ISSUE TITLE
Server v<unknown> located at http://localhost:8088

### ISSUE THREAD

**Issue by robincvlr** [confluentinc/ksql#2126]
Hi everybody, 

I try to use ksql-server and ksql-cli to query my Kafka messages with the following `docker-compose.yml ` : 
```
ksql-server:
    image: confluentinc/cp-ksql-server
    container_name: ksql-server
    depends_on:
      - kafka
    ports:
      - "8088:8088"
    environment:
      KSQL_CONFIG_DIR: "/etc/ksql"
      KSQL_LOG4J_OPTS: "-Dlog4j.configuration=file:/etc/ksql/log4j-rolling.properties"
      KSQL_BOOTSTRAP_SERVERS: "kafka:9092"
      KSQL_APPLICATION_ID: "villo"
      KSQL_LISTENERS: "http://0.0.0.0:8088"
      KSQL_CACHE_MAX_BYTES_BUFFERING: 0
      KSQL_AUTO_OFFSET_RESET: "earliest"
      KSQL_PRODUCER_INTERCEPTOR_CLASSES: "io.confluent.monitoring.clients.interceptor.MonitoringProducerInterceptor"
      KSQL_CONSUMER_INTERCEPTOR_CLASSES: "io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor"
    networks:
      - confluent

  ksql-cli:
    image: confluentinc/cp-ksql-cli
    depends_on:
      - kafka
    environment:
      KSQL_BOOTSTRAP_SERVERS: PLAINTEXT://kafka:9092
      KSQL_LISTENERS: PLAINTEXT://ksql-server:8088
    networks:
      - confluent
    entrypoint: /bin/sh
    tty: true
    stdin_open: true
```
When I attach my ksql-cli docker to launch KSQL, I read the following sentence : 
```
CLI v5.0.0, Server v<unknown> located at http://localhost:8088
Having trouble? Type 'help' (case-insensitive) for a rundown of how things work!
**************** ERROR ********************
Remote server address may not be valid.
Address: http://localhost:8088
Error issuing GET to KSQL server
Caused by: java.net.ConnectException: Connection refused (Connection refused)
Caused by: Could not connect to the server.
*******************************************
```
The `docker logs` command line for ksql-server docker gives me a warning : 
`WARNING: A provider io.confluent.ksql.rest.server.resources.KsqlResource registered in SERVER runtime does not implement any provider interfaces applicable in the SERVER runtime. Due to constraint configuration problems the provider io.confluent.ksql.rest.server.resources.KsqlResource will be ignored. `
Is this related to my issue ? 
I would like to know what goes wrong in my docker composition. I opened my port and succeed my message streaming through the Kafka / Zookeeper / Kafka Connect stack. 
 
Thank you in advance for your interest.

**Comment by tommymsw** [comment]
I also have the same problem , when I set 3 kafka brokers , and start the ksql ,then below error shows:
**************** ERROR ********************
Remote server address may not be valid.
Address: http://localhost:8088
Error issuing GET to KSQL server
Caused by: java.net.ConnectException: Connection refused (Connection refused)
Caused by: Could not connect to the server.
*******************************************

**Comment by deblasis** [comment]
Hi, I had the same issue and by looking around I made it work by unsetting the environment variable **KSQL_CONFIG_DIR**.
Just comment out the relevant line in the `docker-compose.yml` file:
```
#KSQL_CONFIG_DIR: "/etc/ksql"
```

Best regards

**Comment by rmoff** [comment]
If you're using Docker Compose the best pattern to follow for KSQL is to run the server and CLI in the Compose file (e.g. [as shown here](https://github.com/confluentinc/demo-scene/blob/master/community-components-only/docker-compose.yml#L46-L62)) and then invoke the CLI as follows: 

```
docker-compose exec ksql-cli ksql http://ksql-server:8088
```

The key thing is that the KSQL CLI container must be able to access the KSQL Server on the HTTP address given. If you use `localhost:8088` then the CLI will look _locally on the container_.

**Comment by MuqadderIqbal** [comment]
Running into this same error while standing up the confluent-cp-all example and it keeps throwing this same error when connecting to KSQL. The KSQL and CLI containers are up and running, so is Kafka broker. Ports are mapped to defaults and I still can't figure out how to make this thing work. The documentation for this setup does a very poor job of explaining what the "server address" should be and how it affects rest of the components interacting with each other.

**Comment by MuqadderIqbal** [comment]
Never mind. I was able to connect to the KSQL Server container by issuing below command on the ksql-cli prompt:
ksql> server http://ksql-server:8088

**Note:** the KSQL server container name in my case is "ksql-server" and that's why it appears like so in the command above. If you created the container with a different name/port, change the server URL accordingly. Looks like I'm all good now.

**Comment by rmoff** [comment]
Can you provide the link to the documentation you're referring to? We can
then try to improve it :)

**Comment by aspektr** [comment]
Hi, Robin!
I think Muqadder means:
1) https://docs.confluent.io/current/ksql/docs/troubleshoot-ksql.html
2) https://docs.confluent.io/current/ksql/docs/installation/install-ksql-with-docker.html#ksql-cli-connect-to-dockerized-server

I suppose main problem due to we run ksql-server in compose mode and ksql-cli we run via docker run command. We have to cpecify --net=host in such case (if network_mode: host in docker-compose.yml)

example:

`docker run --net=host -it confluentinc/cp-ksql-cli http://0.0.0.0:8088`

It's now obvious from manual






`

**Comment by allenansari174** [comment]
I have an issue running ksql query from python. always I got timeout error:


---------------------------------------------------------------------------
timeout                                   Traceback (most recent call last)
/usr/lib/python3/dist-packages/urllib3/response.py in _error_catcher(self)
    301             try:
--> 302                 yield
    303 

/usr/lib/python3/dist-packages/urllib3/response.py in read_chunked(self, amt, decode_content)
    600                     break
--> 601                 chunk = self._handle_chunk(amt)
    602                 decoded = self._decode(chunk, decode_content=decode_content,

/usr/lib/python3/dist-packages/urllib3/response.py in _handle_chunk(self, amt)
    566             returned_chunk = self._fp._safe_read(self.chunk_left)
--> 567             self._fp._safe_read(2)  # Toss the CRLF at the end of the chunk.
    568             self.chunk_left = None

/usr/lib/python3.6/http/client.py in _safe_read(self, amt)
    611         while amt > 0:
--> 612             chunk = self.fp.read(min(amt, MAXAMOUNT))
    613             if not chunk:

/usr/lib/python3.6/socket.py in readinto(self, b)
    585             try:
--> 586                 return self._sock.recv_into(b)
    587             except timeout:

timeout: timed out

During handling of the above exception, another exception occurred:

ReadTimeoutError                          Traceback (most recent call last)
/usr/lib/python3/dist-packages/requests/models.py in generate()
    744                 try:
--> 745                     for chunk in self.raw.stream(chunk_size, decode_content=True):
    746                         yield chunk

/usr/lib/python3/dist-packages/urllib3/response.py in stream(self, amt, decode_content)
    431         if self.chunked and self.supports_chunked_reads():
--> 432             for line in self.read_chunked(amt, decode_content=decode_content):
    433                 yield line

/usr/lib/python3/dist-packages/urllib3/response.py in read_chunked(self, amt, decode_content)
    625             if self._original_response:
--> 626                 self._original_response.close()

/usr/lib/python3.6/contextlib.py in __exit__(self, type, value, traceback)
     98             try:
---> 99                 self.gen.throw(type, value, traceback)
    100             except StopIteration as exc:

/usr/lib/python3/dist-packages/urllib3/response.py in _error_catcher(self)
    306                 # there is yet no clean way to get at it from this context.
--> 307                 raise ReadTimeoutError(self._pool, None, 'Read timed out.')
    308 

ReadTimeoutError: HTTPConnectionPool(host='localhost', port=8088): Read timed out.

During handling of the above exception, another exception occurred:

ConnectionError                           Traceback (most recent call last)
<ipython-input-2-2426a0b79759> in <module>
      1 query = client.query('select * from user_tweet_count')
----> 2 for item in query: print(item)

/usr/local/lib/python3.6/dist-packages/ksql/api.py in query(self, query_string, encoding, chunk_size, stream_properties, idle_timeout)
     69         streaming_response = self._request(endpoint='query', sql_string=query_string, stream_properties=stream_properties)
     70         start_idle = None
---> 71         for chunk in streaming_response.iter_content(chunk_size=chunk_size):
     72             if chunk != b'\n':
     73                 start_idle = None

/usr/lib/python3/dist-packages/requests/models.py in generate()
    750                     raise ContentDecodingError(e)
    751                 except ReadTimeoutError as e:
--> 752                     raise ConnectionError(e)
    753             else:
    754                 # Standard file-like object.

ConnectionError: HTTPConnectionPool(host='localhost', port=8088): Read timed out.

**Comment by setlaoui** [comment]
try this:  docker exec -it ksqldb-server ksql http://ksqldb-server:8088

**Comment by atulrenapurkar** [comment]
yes, this worked because in docker-compose.yml file contain hostname: ksqldb-server

**Comment by kelvin-chang-kc9457** [comment]
I had the same problem. Sharing how it was resolved. 

The following did not work for me:
docker exec -it test_demo_ksqldb-cli ksql http://test_demo_ksqldb-server:8088

While this did:
docker exec -it test-demo-ksqldb-cli ksql http://test-demo-ksqldb-server:8088

It seems the underscore in the domain name is causing the problem.

I searched on google and apparently underscores are not allowed in domain names.
